### ISSUE TITLE
ArrowFileReader.getRecordBlocks() seems incompatibility with JDK 8

### ISSUE THREAD

**Issue by ovasty** [apache/arrow#44311]
### Describe the bug, including details regarding any error messages, version, and platform.

I run Arrow's demo codes under JDK8 and Arrow 17.0.0, but it seems not work.
IPC reader example code as below:
`try(
                BufferAllocator allocator = new RootAllocator(Long.MAX_VALUE);
                FileInputStream fileInputStream = new FileInputStream(new File("random_access_file.arrow"));
                ArrowFileReader reader = new ArrowFileReader(fileInputStream.getChannel(), allocator);
        ){
            System.out.println("Record batches in file: " + reader.getRecordBlocks().size());
            for (ArrowBlock arrowBlock : reader.getRecordBlocks()) {
                reader.loadRecordBatch(arrowBlock);
                VectorSchemaRoot root = reader.getVectorSchemaRoot();
                System.out.println("VectorSchemaRoot read: \n" + root.contentToTSVString());
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
`

demo
maven:
`
<dependency>
            <groupId>org.apache.arrow</groupId>
            <artifactId>arrow-vector</artifactId>
            <version>17.0.0</version>
        </dependency>
        <dependency>
            <groupId>org.apache.arrow</groupId>
            <artifactId>arrow-memory-netty</artifactId>
            <version>17.0.0</version>
        </dependency>
`
the error info is:
`
Exception in thread "main" java.lang.NoSuchMethodError: java.nio.ByteBuffer.flip()Ljava/nio/ByteBuffer;
	at org.apache.arrow.vector.ipc.ArrowFileReader.readSchema(ArrowFileReader.java:94)
	at org.apache.arrow.vector.ipc.ArrowReader.initialize(ArrowReader.java:179)
	at org.apache.arrow.vector.ipc.ArrowFileReader.initialize(ArrowFileReader.java:121)
	at org.apache.arrow.vector.ipc.ArrowReader.ensureInitialized(ArrowReader.java:172)
	at org.apache.arrow.vector.ipc.ArrowFileReader.getRecordBlocks(ArrowFileReader.java:180)
	at com.hd.arrowdemo.Main.ipcReader(Main.java:93)
	at com.hd.arrowdemo.Main.main(Main.java:44)
`

### Component(s)

Java

**Comment by vibhatha** [comment]
Just curious, what happens if you run with JDK 11?

**Comment by saurabhchauhan006** [comment]
So the problem is that arrow 17.0.0 version is compiled using java 9 and above.
In ArrowFileReader.readSchema() it is calling ByteBuffer.flip() and because in java 8 there is no implementation for ByteBuffer.flip() it fails.
To fix this I have forked this class and type casted ByteBuffer class with Buffer.
`@Override
    protected Schema readSchema() throws IOException {
        if (footer == null) {
            if (in.size() <= (ArrowMagic.MAGIC_LENGTH * 2 + 4)) {
                throw new InvalidArrowFileException("file too small: " + in.size());
            }
            ByteBuffer buffer = ByteBuffer.allocate(4 + ArrowMagic.MAGIC_LENGTH);
            long footerLengthOffset = in.size() - buffer.remaining();
            in.setPosition(footerLengthOffset);
            in.readFully(buffer);
            ((Buffer)buffer).flip();
            byte[] array = buffer.array();
            if (!ArrowMagic.validateMagic(Arrays.copyOfRange(array, 4, array.length))) {
                throw new InvalidArrowFileException(
                        "missing Magic number " + Arrays.toString(buffer.array()));
            }
            int footerLength = MessageSerializer.bytesToInt(array);
            if (footerLength <= 0
                    || footerLength + ArrowMagic.MAGIC_LENGTH * 2 + 4 > in.size()
                    || footerLength > footerLengthOffset) {
                throw new InvalidArrowFileException("invalid footer length: " + footerLength);
            }
            long footerOffset = footerLengthOffset - footerLength;
            LOGGER.debug("Footer starts at {}, length: {}", footerOffset, footerLength);
            ByteBuffer footerBuffer = ByteBuffer.allocate(footerLength);
            in.setPosition(footerOffset);
            in.readFully(footerBuffer);
            footerBuffer.flip();
            Footer footerFB = Footer.getRootAsFooter(footerBuffer);
            this.footer = new ArrowFooter(footerFB);
        }
        MetadataV4UnionChecker.checkRead(footer.getSchema(), footer.getMetadataVersion());
        return footer.getSchema();
    }`

So basically in java 9 and above ByteBuffer.flip() just calls Buffer.flip(). So by this way it solves the problem with Java 8.
I am not pushing a fix for this as the current version of Arrow requires java 11 and above, so there is no point of fixing this now.
